src/Microsoft.AspNetCore.Hosting/Startup/StartupBase.cs

```diff
@@ -16,4 +16,25 @@ public virtual IServiceProvider ConfigureServices(IServiceCollection services)
             return services.BuildServiceProvider();
         }
     }
+
+    public abstract class StartupBase<TContainer> : IStartup
+    {
+        private readonly IServiceProviderFactory _factory;
+
+        public StartupBase(IServiceProviderFactory factory)
+        {
+            _factory = factory;
+        }
+
+        public abstract void Configure(IApplicationBuilder app);
+
+        public IServiceProvider ConfigureServices(IServiceCollection services)
+        {
+            var builder = _factory.CreateBuilder(services);
+            ConfigureContainer((TContainer)builder);
```

Comment:
> The fact that you have to flow the factory to the base feels like a deal breaker.

Yeah, it sucks, but I'd still use the base class.

The only way I can think of, is to add it as a parameter to `IStartup.ConfigureServices`. You could easily "hide it" from the average user by having abstract classes that deals with it. It would break `IStartup` though :cry: